### PR TITLE
added some warnings and tests

### LINKED ISSUE #3088: Modules V2 - Warnings and Helpers
Ensure the following warnings are in place, with helpful messages and the corresponding limitation documented

- [x] Shadow warnings for venv/libs 
- [x] A V2 module discovered via git, that is not installed: urge the user to install it and fail
- [x] When finding a valid namespace package without cf files, produce a warning to no use a normal `pip install` or `pip build` on modules.

### LINKED ISSUE #3178: moduletool package warns about unpackaged pyc files
The `moduletool package` command warns about any files that were not included in the package. It should exclude `*.pyc` files from this warning. Here's an example warning:
```
WARNING  inmanta.moduletool:moduletool.py:755 The following files are present in the inmanta_plugins/minimalv2module directory on disk, but were not packaged: ['__pycache__/__init__.cpython-39.pyc']. Update you setup.cfg file if they need to be packaged.
```

### LINKED ISSUE #3179: v2 modules warning that name doesn't match directory
```
WARNING  inmanta.module:module.py:1376 The name in the module file (inmanta-module-minimalv2module) does not match the directory name (minimalv2module)
```
This is obviously not relevant for v2 modules.

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Added some warnings, improved some exception messages, added tests for both.

closes #3088 
closes #3178 
closes #3179

## output
Because the exceptions are the focus of this PR are stacked by `__cause__` the actual command output might be hard to follow from the test cases alone. Therefor I'm providing an overview of what the end user would see here.

### 1. load v2 from v1 libs
```
Failed to load module minimalv2module (reported in import minimalv2module (./main.cf:1))
caused by:
  Could not load module minimalv2module
  caused by:
    inmanta.module.ModuleMetadataFileNotFound: Module at /home/sander/documents/projects/inmanta/inmanta-model-test/libs/minimalv2module looks like a v2 module. Please add it as a v2 requirement with `inmanta modules add <module_name>`
```

### 2. incorrectly installed v2 module (`pip install` instead of `inmanta module install`
- case 1: `setup.cfg` missing:
```
Failed to load module minimalv2module (reported in import minimalv2module (./main.cf:1))
caused by:
  Could not load module minimalv2module
  caused by:
    inmanta.module.InvalidModuleException: Invalid module: found module package but it has no setup.cfg. This occurs when you install or build modules from source incorrectly. Always use the `inmanta module` `install` and `build` commands to respectively install and build modules from source.
```
- case 2: no model:
```
Failed to load module minimalv2module (reported in import minimalv2module (./main.cf:1))
caused by:
  Could not load module minimalv2module
  caused by:
    inmanta.module.InvalidModuleException: The module at /home/sander/.virtualenvs/inmanta-core/lib/python3.9/site-packages/inmanta_plugins/minimalv2module contains no model (.cf) files. This occurs when you install or build modules from source incorrectly. Always use the `inmanta module` `install and `build` commands to respectively install and build modules from source.
```

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/env.py [hunk 1]
```diff
@@ -173,7 +173,9 @@ def _run_command_and_log_output(
         try:
             output = subprocess.check_output(cmd, stderr=stderr, env=env)
         except CalledProcessError as e:
-            LOGGER.error("%s: %s", cmd, e.output.decode())
+            err: str = e.stderr.decode()
+            msg: str = err if err else e.output.decode()
+            LOGGER.error("%s: %s", cmd, msg)
```
Comment (sanderr) [inline src/inmanta/env.py:178]: This has nothing to do with the warnings, I just noticed for the nth time that this error message wasn't all that useful to determine the actual cause so I thought I'd make an attempt to fix it.

File: tests/test_modules.py [hunk 2]
```diff
@@ -206,7 +222,7 @@ def test_get_requirements(
 
 @pytest.mark.parametrize("editable", [True, False])
 def test_module_v2_source_get_installed_module_editable(
-    snippetcompiler,
+    snippetcompiler_clean,
```
Comment (sanderr) [inline tests/test_modules.py:225]: This test installs test packages in the snippetcompiler venv. Not using a clean snippetcompiler causes instability issues for other tests.

File: src/inmanta/env.py (hunk 1 above)
Comment (wouterdb) [reply src/inmanta/env.py:178]: Nice!

File: src/inmanta/module.py [hunk 3]
```diff
@@ -2300,6 +2297,20 @@ def __init__(self, project: Optional[Project], path: str):
             super(ModuleV1, self).__init__(project, path)
         except InvalidMetadata as e:
             raise InvalidModuleException(f"The module found at {path} is not a valid V1 module") from e
+        except ModuleMetadataFileNotFound:
+            if os.path.exists(os.path.join(path, ModuleV2.MODULE_FILE)):
+                raise ModuleMetadataFileNotFound(
+                    f"Module at {path} looks like a v2 module. Please add it as a v2 requirement with"
+                    " `inmanta modules add <module_name>`"
```
Comment (wouterdb) [inline src/inmanta/module.py:2304]: @arnaudsjs is this correct or does it require a `--v2`

File: tests/test_modules.py (hunk 2 above)
Comment (wouterdb) [reply tests/test_modules.py:225]: I would add that comment to the code

File: tests/test_modules.py [hunk 4]
```diff
@@ -249,3 +265,62 @@ def test_module_v2_source_path_for_v1(snippetcompiler) -> None:
...
+    found in a v1 module source.
+    """
+    with pytest.raises(module.ModuleLoadingException) as excinfo:
+        snippetcompiler.setup_for_snippet("import minimalv2module", add_to_module_path=[modules_v2_dir])
+    cause: CompilerException = excinfo.value.__cause__
+    root_cause: Optional[BaseException] = cause.__cause__
+    assert root_cause is not None
+    assert isinstance(root_cause, module.InvalidModuleException)
+    assert root_cause.msg == (
+        "Module at %s looks like a v2 module. Please add it as a v2 requirement with `inmanta modules add <module_name>`"
```
Comment (wouterdb) [inline tests/test_modules.py:282]: perhaps it would be good to also extend the testcase to include the fix proposed by the error message?

(I also wonder in this case if inmanta modules add is going to help at all?)

File: src/inmanta/module.py (hunk 3 above)
Comment (sanderr) [reply src/inmanta/module.py:2304]: Good point, I don't think this is correct.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/moduletool/test_build.py [hunk 5]
```diff
@@ -164,6 +164,12 @@ def test_build_v2_module_incomplete_package_data(tmpdir, modules_v2_dir: str, ca
         )
     )
 
+    # write some garbage cfcache and pyc files to verify those are ignored as well
+    open(os.path.join(source_dir, "test.pyc"), "w").close()
+    cfcache_dir: str = os.path.join(module_copy_dir, "model", "__cfcache__")
+    os.makedirs(cfcache_dir, exist_ok=True)
+    open(os.path.join(cfcache_dir, "test.cfc"), "w").close()
+
```
[the remark this answers is not in the document]
Comment (sanderr) [reply tests/moduletool/test_build.py:172]: The assertion in the next part checks the warning message.

File: tests/test_modules.py [hunk 6]
```diff
@@ -69,6 +73,18 @@ def test_wrong_name(self):
...
 
+    def test_non_matching_name_v2_module(self) -> None:
+        """
+        Make sure the warning regarding directory name does not trigger for v2 modules, as it is not relevant there.
+        """
+        template_dir: str = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data", "modules_v2", "minimalv2module")
+        with tempfile.TemporaryDirectory() as tmpdir:
+            mod_dir: str = os.path.join(tmpdir, "not-the-module-name")
+            module_from_template(template_dir, mod_dir)
+            module.ModuleV2(project=mock.Mock(), path=mod_dir)
```
[the remark this answers is not in the document]
Comment (sanderr) [reply tests/test_modules.py:84]: I copied this from the other test case in this class. `DummyProject` would work just as well. I'll update it.

File: tests/moduletool/test_build.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply tests/moduletool/test_build.py:172]: True. No clue how I missed that.

File: tests/test_modules.py (hunk 4 above)
Comment (sanderr) [reply tests/test_modules.py:282]: It does. Why would you think it wouldn't?
I've extended both testcases though the second one is still failing because of #3321.

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/module.py [hunk 7]
```diff
@@ -2301,7 +2301,7 @@ def __init__(self, project: Optional[Project], path: str):
             if os.path.exists(os.path.join(path, ModuleV2.MODULE_FILE)):
                 raise ModuleMetadataFileNotFound(
                     f"Module at {path} looks like a v2 module. Please add it as a v2 requirement with"
-                    " `inmanta modules add <module_name>`"
+                    " `inmanta module add --v2 <module_name>`"
```
Comment (wouterdb) [inline src/inmanta/module.py:2304]: I wonder, should we just put the correct command in the message? (replace `<module_name>` with the actual module name?_

File: src/inmanta/module.py [hunk 8]
```diff
@@ -2497,7 +2497,7 @@ def __init__(
 
         if not os.path.exists(os.path.join(self.model_dir, "_init.cf")):
             raise InvalidModuleException(
-                f"The module at {path} contains no model (.cf) files. This occurs when you install or build modules from source"
+                f"The module at {path} contains no _init.cf file. This occurs when you install or build modules from source"
                 " incorrectly. Always use the `inmanta module` `install and `build` commands to respectively install and build"
```
Comment (wouterdb) [inline src/inmanta/module.py:2501]: I would make this more copy-past friendly by writing out both command in full.

File: tests/test_modules.py [hunk 9]
```diff
@@ -267,24 +269,54 @@ def test_module_v2_source_path_for_v1(snippetcompiler) -> None:
...
+        install_project=False,
+    )
+    os.chdir(project.path)
+    project_init = module.Project.__init__
+
+    # patch Project.__init__ to set autostd=False because v2 std does not exist yet
+    def project_init_nostd(self, *args, **kwargs) -> None:
+        project_init(self, *args, autostd=False, **{key: value for key, value in kwargs.items() if key != "autostd"})
+
+    with mock.patch.object(module.Project, "__init__", project_init_nostd):
```
Comment (wouterdb) [inline tests/test_modules.py:302]: in pytest I think using monkeypatching is the way to go?

**Comment by sanderr** [comment]
Improved exception format?
```
Failed to load module minimalv2module (reported in import minimalv2module (./main.cf:1))
caused by:
  Module at /home/sander/documents/projects/inmanta/inmanta-model-test/libs/minimalv2module looks like a v2 module. Please add it as a v2 requirement with `inmanta module add --v2 <module_name>`
```

File: tests/test_modules.py (hunk 4 above)
Comment (wouterdb) [reply tests/test_modules.py]: because it doesn't work if the v2 module is not in a configured pip index.

**Comment by sanderr** [comment]
Note to self: make sure exception messages are accurate. Create ticket to make them more specific with explainer.

**Comment by sanderr** [comment]
I've got some review work to do so I've ticketized the explainer after all: #3325.

**Review by wouterdb — APPROVED** [review]
